tls/s2n_renegotiate.c

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tls/s2n_renegotiate.h"
+
+#include "error/s2n_errno.h"
+#include "stuffer/s2n_stuffer.h"
+#include "tls/s2n_connection.h"
+#include "utils/s2n_safety.h"
+
+/*
+ * Prepare a connection to be reused for a second handshake.
+ *
+ * s2n-tls was not originally designed to support renegotiation.
+ * s2n_connection is a very large structure with configuration fields set by the application
+ * mixed in with internal state fields set by the handshake.
+ * Ensuring all existing internal state fields (and any new fields added) are safe to reuse for
+ * a renegotiated handshake would be extremely prone to errors.
+ * For safety, we instead wipe the entire connection and only restore fields we know we need
+ * in order to continue sending and receiving encrypted data.
+ *
+ * Any configuration fields set by the application will need to be set by the application again.
+ */
+int s2n_renegotiate_wipe(struct s2n_connection *conn)
+{
+    POSIX_ENSURE_REF(conn);
+
+    /* We use this method to reset both clients and servers when testing.
+     * However, outside of tests, it should only be called for client connections
+     * because we only support renegotiation for clients.
+     */
+    POSIX_ENSURE(conn->mode == S2N_CLIENT || s2n_in_unit_test(), S2N_ERR_NO_RENEGOTIATION);
+
+    /* Best effort check for pending input or output data.
+     * This method should not be called until the application has stopped sending and receiving.
+     * Saving partial read or parital write state would complicate this problem.
+     */
+    POSIX_ENSURE(s2n_stuffer_data_available(&conn->header_in) == 0, S2N_ERR_INVALID_STATE);
+    POSIX_ENSURE(s2n_stuffer_data_available(&conn->in) == 0, S2N_ERR_INVALID_STATE);
```

Comment:
It's not like crazy hard, but they do need to check s2n_peek(). I think it's harder than s2n_send() because a successful s2n_recv() call doesn't automatically mean that the IO buffers are empty.

Reply:
I'm considering not supporting this initially, but adding it after the MVP is done to improve the user experience.